File: Zad5.java
// Abstrakcja usługi powiadamiania
interface MessageService {
    void sendMessage(String message);
}

// Konkretną implementacja usługi - Email
class EmailService implements MessageService {
    @Override
    public void sendMessage(String message) {
        System.out.println("Sending email: " + message);
    }
}

// Klasa powiadomień
class Notification {

    private MessageService messageService;

    // Konstruktor zależny od abstrakcji
    public Notification(MessageService messageService) {
        this.messageService = messageService;
    }

    public void notify(String message) {
        messageService.sendMessage(message);
    }
}

public class Zad5 {
    public static void main(String[] args) {
        // Używamy abstrakcji do tworzenia obiektów powiadomień
        MessageService emailService = new EmailService();
        Notification notification = new Notification(emailService);
        
        // Testowanie systemu powiadomień
        notification.notify("Powiadomienie z Politechniki Poznanskiej");
    }
}

File: zad5.cc
#include <iostream>
#include <string>

// Abstrakcja usługi powiadomień
class MessageService {
public:
    virtual void sendMessage(const std::string& message) = 0;
};

// Konkretną implementacja usługi - Email
class EmailService : public MessageService {
public:
    void sendMessage(const std::string& message) override {
        std::cout << "Sending email: " << message << std::endl;
    }
};

// Klasa powiadomień
class Notification {
private:
    MessageService* messageService;

public:
    // Konstruktor zależny od abstrakcji (MessageService)
    Notification(MessageService* service) : messageService(service) {}

    void notify(const std::string& message) {
        messageService->sendMessage(message);
    }
};

int main() {
    // Tworzymy konkretną implementację usługi
    EmailService emailService;
    
    // Tworzymy obiekt powiadomień, który zależy od abstrakcji
    Notification notification(&emailService);
    
    // Testowanie systemu powiadomień
    notification.notify("Witamy Na Politechnice Poznańskiej");

    return 0;
}

